### DISCUSSION TITLE
Conditional Creation of Environment Variable

### DISCUSSION THREAD

**Discussion by sansa20** [wixtoolset/issues#7272] [Questions]
I have a setup project for creating an msi installer. But one of my pre-requisites for the installer is that a specific environment variable should be present in the system. So first I need to check if the environment variable exists. If it does then the installation will proceed, and if not, the environment variable will be automatically created and then the installation will happen.

To implement this I am first using the "RegistrySearch" element to check whether the environment variable is present. This is then followed by a "CustomAction" where I define the value of the environment variable to be created. My question is whether it is right to use a "RegistrySearch" to check for an environment variable? I have found some sources which say it is okay while others say it is inaccurate. Following is my code for reference

       <Property Id="TAPPATHEXISTS">
		<RegistrySearch Id="CheckEnvVar"
						Root="HKCU"
						Key="Environment"
						Name="TAP_PATH"
						Type="raw"/>
	</Property>
	<CustomAction Id="SetTapPathVariable" Property="SetEnvironmentVar" Value="cmd /c setx TAP_PATH C:\Program Files (x86)\Keysight\Test Automation" />

	<InstallExecuteSequence>
		<Custom Action="SetTapPathVariable" Before="InstallFiles">
			<![CDATA[MY_PROPERTY = "TRUE"]]>  
		</Custom>

**Comment by robmen** [comment]
I think if you use the `Environment` element, everything will work properly. I happen to have a [Deployment Dojo episode about this very thing](https://youtube.com/live/0ApAkl4HKxw?feature=share).

**Reply by sansa20** [reply]
I have referred to the deployment dojo episode you mentioned above and applied the same for my creation of an environment variable.
I am successfully able to build my project and run the application as well. My dlls are stored in the defined install location as well. However, the environment variable is still not being created. I am unable to figure out what is going wrong as the code seems to be fine. I am attaching the relevant lines of code that I have used.

```
<Fragment>
		<ComponentGroup Id="EnvVar" Directory="INSTALLFOLDER">
			<Component>
				<File Source="OpenTapWixSetup.msi" />
				<Environment Id="TapPathVar" Name="TAP_PATH" System="yes" Value="C:\Program Files (x86)\Keysight\Test Automation" Action="set" Part="all"/>
			</Component>
		</ComponentGroup>
	</Fragment>
```

**Reply by jhennessey** [reply]
I'd recommend [logging](https://wixtoolset.org/docs/v3/howtos/general/get_a_log/) your installation to verify the component is actually being installed. If it isn't, the environment variable won't get set.

**Comment by chrpai** [comment]
At the risk of offending, and I'm sorry if I do, this has a certain code smell to it.  I've never heard of an installer requiring an environment variable.  That sounds like trying to bypass passing properties into the installer.  And when you say "it must exist"... where?  In per-user?  In per-machine?  For the process that is running client side or a new process not yet running or for a service already running?

If you really want to just get it you can use [%ENVIRONMENTNAME] an AppSearch shouldn't be needed.   But beware, you can set a machine wide environment variable and processes started by the Service Control Manger won't get them until you've rebooted.  This includes the server side of msiexec that runs during a silent install.

MSI is fairly decent about authoring environment variables but I can think of one really huge gap.  There's a tiny note in the SDK doco that says MSI doesn't send a WM_SETTINGCHANGE message when a reboot is pending.  Personally I think this is nuts but that's the behavior.  WiX has some goodness for this.

https://learn.microsoft.com/en-us/windows/win32/msi/conditional-statement-syntax

https://learn.microsoft.com/en-us/windows/win32/msi/writeenvironmentstrings-action

https://wixtoolset.org/docs/v3/customactions/wixsettingchange/
